File: examples/python/holesail_server.py
#!/usr/bin/env python3
"""
holesail-py -- Expose a local TCP port over HyperDHT.

A Python reimplementation of holesail's server functionality,
powered by hyperdht-cpp. Accepts encrypted P2P connections and
bridges them to a local TCP port.

Usage:
    python holesail_server.py --live <port> [options]

Options:
    --live <port>     Local TCP port to expose (required)
    --host <addr>     Local bind address (default: 127.0.0.1)
    --seed <hex>      64-char hex seed for deterministic identity
    --secure          Firewall rejects peers that don't know the seed

Examples:
    # Expose a local web server
    python holesail_server.py --live 8080

    # Deterministic identity (survives restarts)
    python holesail_server.py --live 3000 --seed $(head -c32 /dev/urandom | xxd -p -c64)

    # Secure mode (only seed-holders can connect)
    python holesail_server.py --live 8080 --seed abc123...def --secure

Connect with JS holesail:
    holesail --connect <connection_string>
"""

from __future__ import annotations

import argparse
import base64
import os
import socket
import sys
import weakref

sys.path.insert(0, ".")

from hyperdht import HyperDHT, KeyPair

# z32 encoding (same alphabet as holesail's z32 module)
_Z32 = "ybndrfg8ejkmcpqxot1uwisza345h769"
_STD = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"


def _to_z32(data: bytes) -> str:
    b32 = base64.b32encode(data).decode().rstrip("=")
    return b32.translate(str.maketrans(_STD, _Z32.upper())).lower()


def _parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        prog="holesail-py",
        description="Expose a local TCP port over HyperDHT (P2P)",
    )
    parser.add_argument(
        "--live", type=int, required=True, metavar="PORT",
        help="Local TCP port to expose",
    )
    parser.add_argument(
        "--host", default="127.0.0.1",
        help="Local bind address (default: 127.0.0.1)",
    )
    parser.add_argument(
        "--seed", metavar="HEX",
        help="64-char hex seed for deterministic keypair",
    )
    parser.add_argument(
        "--secure", action="store_true",
        help="Reject peers that don't know the seed",
    )
    return parser.parse_args()


def _make_keypair(
    seed_hex: str | None, secure: bool,
) -> tuple[KeyPair, str]:
    """Create keypair and connection string."""
    if seed_hex:
        if len(seed_hex) != 64:
            print("Error: --seed must be 64 hex characters (32 bytes)")
            sys.exit(1)
        seed = bytes.fromhex(seed_hex)
        kp = KeyPair.from_seed(seed)
        connection_key = _to_z32(seed if secure else kp.public_key)
    elif secure:
        raw_seed = os.urandom(32)
        kp = KeyPair.from_seed(raw_seed)
        connection_key = _to_z32(raw_seed)
    else:
        kp = KeyPair.generate()
        connection_key = _to_z32(kp.public_key)

    return kp, connection_key


class Bridge:
    """Bidirectional bridge between a TCP socket and an encrypted stream.

    All I/O runs on the libuv event loop — TCP sockets are registered
    via uv_poll (hyperdht_poll_start), so there's zero polling latency.
    The stream->TCP direction is handled by the stream's on_data callback.
    The TCP->stream direction fires when libuv detects the fd is readable.
    """

    def __init__(
        self, tcp_sock: socket.socket, stream, dht,
        bridges: set[Bridge],
    ) -> None:
        self._tcp = tcp_sock
        self._stream = stream
        self._dht = dht
        self._bridges = bridges
        self._poll_handle = None
        self._write_buf = bytearray()
        self._watching_writable = False
        self._closed = False
        self._tcp.setblocking(False)

    def activate(self) -> None:
        """Start watching TCP for data. Call from on_open."""
        if self._poll_handle or self._closed:
            return
        # Single poll handle — libuv forbids multiple polls on same fd.
        # Start with readable only; add writable when backpressured.
        self._poll_handle = self._dht.poll_start(
            self._tcp.fileno(), self._on_poll)

    def _on_poll(self, fd: int, events: int) -> None:
        """Single callback for both readable and writable events."""
        if events & 1:  # POLL_READABLE
            self._handle_readable()
        if events & 2:  # POLL_WRITABLE
            self._handle_writable()

    def _handle_readable(self) -> None:
        if not self._stream:
            return
        try:
            data = self._tcp.recv(65536)
        except BlockingIOError:
            return
        except OSError:
            self.close()
            return

        if not data:
            self.close()
            return

        try:
            self._stream.write(data)
        except RuntimeError:
            self.close()

    def _handle_writable(self) -> None:
        if not self._write_buf:
            self._set_writable(False)
            return

        try:
            sent = self._tcp.send(self._write_buf)
        except BlockingIOError:
            return
        except OSError:
            self.close()
            return

        del self._write_buf[:sent]

        if not self._write_buf:
            self._set_writable(False)

    def write_to_tcp(self, data: bytes) -> None:
        """Called from stream's on_data — forward to TCP."""
        if self._closed:
            return

        # If we already have buffered data, append and let the
        # writable poll flush it in order.
        if self._write_buf:
            self._write_buf.extend(data)
            return

        # Try to send directly first (fast path).
        try:
            sent = self._tcp.send(data)
        except BlockingIOError:
            sent = 0
        except OSError:
            self.close()
            return

        if sent == len(data):
            return  # All sent, no buffering needed

        # Partial send — buffer the rest and watch for writable.
        self._write_buf.extend(data[sent:])
        self._set_writable(True)

    def _set_writable(self, enable: bool) -> None:
        """Toggle writable event on the single poll handle."""
        if enable == self._watching_writable:
            return
        self._watching_writable = enable
        if not self._poll_handle:
            return
        # Recreate poll with updated event flags — libuv doesn't
        # support modifying events on an active poll, so stop+restart.
        self._dht.poll_stop(self._poll_handle)
        self._poll_handle = self._dht.poll_start(
            self._tcp.fileno(), self._on_poll,
            readable=True, writable=enable)

    def close(self) -> None:
        if self._closed:
            return
        self._closed = True
        self._bridges.discard(self)
        if self._poll_handle:
            self._dht.poll_stop(self._poll_handle)
            self._poll_handle = None
        self._write_buf.clear()
        try:
            self._tcp.close()
        except OSError:
            pass
        if self._stream:
            try:
                self._stream.close()
            except RuntimeError:
                pass
            self._stream = None


def main() -> None:
    args = _parse_args()
    local_port = args.live
    local_host = args.host

    # Verify the local port is reachable before joining the DHT
    test_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    try:
        test_sock.connect((local_host, local_port))
    except OSError:
        print(f"Warning: nothing listening on {local_host}:{local_port} yet")
    finally:
        test_sock.close()

    # Keypair + connection string
    kp, connection_key = _make_keypair(args.seed, args.secure)

    # Build the DHT node
    dht = HyperDHT(use_public_bootstrap=True)
    dht.bind()

    bridges: set[Bridge] = set()

    # Create server with optional firewall
    server = dht.create_server()
    server.set_reusable_socket(True)  # skip holepunch on reconnect from same peer

    if args.secure:
        server.set_firewall(
            lambda remote_pk, host, port: remote_pk != kp.public_key
        )

    connection_count = 0

    def on_connection(conn) -> None:
        nonlocal connection_count
        connection_count += 1
        print(f"  Peer connected from {conn.peer_host}:{conn.peer_port}"
              f"  [{connection_count} total]")

        # Connect to local TCP server
        tcp_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        try:
            tcp_sock.connect((local_host, local_port))
        except OSError as e:
            print(f"  Error: can't reach {local_host}:{local_port} -- {e}")
            tcp_sock.close()
            return

        # Create bridge with stream=None initially, set after open_stream
        bridge = Bridge(tcp_sock, None, dht, bridges)
        bridges.add(bridge)

        conn_id = connection_count

        def on_open() -> None:
            bridge.activate()

        def on_data(data: bytes) -> None:
            bridge.write_to_tcp(data)

        def on_close() -> None:
            print(f"  Stream closed  [{conn_id}]")
            bridge.close()

        stream = dht.open_stream(
            conn, on_open=on_open, on_data=on_data, on_close=on_close,
        )
        bridge._stream = stream

    server.listen(kp, on_connection)

    # Publish metadata so holesail clients can discover port/protocol.
    # JS holesail does: dht.mutablePut(keyPair, JSON({host, port, udp}))
    import json
    metadata = json.dumps({
        "host": local_host,
        "port": local_port,
        "udp": False,
    }).encode()
    dht.mutable_put(kp, metadata, seq=1)

    # Connection string
    prefix = "s000" if args.secure else "0000"
    hs_link = f"hs://{prefix}{connection_key}"

    print(f"""
  holesail-py -- P2P tunnel powered by hyperdht-cpp
  --------------------------------------------------

  Exposing:    {local_host}:{local_port}
  DHT port:    {dht.port}
  Mode:        {"secure" if args.secure else "open"}
  Public key:  {kp.public_key.hex()}

  Connection string:
    {hs_link}

  Connect with:
    holesail --connect {hs_link}

  Ctrl+C to stop
""")

    # Diagnostic heartbeat — print DHT state every 30s
    import time
    last_heartbeat = [time.monotonic()]

    orig_run = dht.run

    def run_with_heartbeat():
        while True:
            dht.run(mode="once")
            now = time.monotonic()
            if now - last_heartbeat[0] >= 30:
                last_heartbeat[0] = now
                addr = dht.remote_address
                addr_str = f"{addr.host}:{addr.port}" if addr else "unknown"
                print(f"  [heartbeat] online={dht.is_online} "
                      f"persistent={dht.is_persistent} "
                      f"listening={server.is_listening} "
                      f"addr={addr_str} "
                      f"bridges={len(bridges)} "
                      f"conns={connection_count}")

    # Event-driven: TCP sockets are registered with libuv via poll_start,
    # so dht.run() handles everything — DHT, streams, and TCP bridging.
    try:
        run_with_heartbeat()
    except KeyboardInterrupt:
        pass

    print(f"\n  Shutting down ({connection_count} connections served)")
    for bridge in list(bridges):
        bridge.close()
    server.close_force()  # skip UNANNOUNCE — exit fast
    dht.destroy(force=True)


if __name__ == "__main__":
    main()

File: examples/rust/src/holesail_server.rs
//! holesail-rs (server side) — expose a local TCP port over HyperDHT.
//!
//! A Rust reimplementation of holesail's server functionality, mirroring
//! `examples/python/holesail_server.py`. Accepts encrypted P2P
//! connections from the public DHT and bridges each one to a local
//! TCP service.
//!
//! Usage:
//!   holesail-server --live 8080
//!   holesail-server --live 3000 --seed <64-hex>
//!   holesail-server --live 8080 --seed <64-hex> --secure
//!
//! Connect with the JS holesail client (or any peer that has our
//! connection string):
//!   holesail --connect hs://<prefix><z32-of-key>
//!
//! Why this is simpler than the Python version: tokio gives us native
//! async TCP, so we skip the `uv_poll`-based hand-rolled bridge and
//! just spawn `tokio::io::copy_bidirectional(&mut tcp, &mut stream)`
//! per accepted peer. ~80 lines of bridging in Python collapses to
//! ~10 lines here.

use std::net::SocketAddr;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Arc;
use std::time::Duration;

use clap::Parser;
use hyperdht::{Dht, DhtOptions, Keypair, ServerOptions};
use tokio::net::TcpStream;

#[derive(Parser, Debug)]
#[command(name = "holesail-server", about = "Expose a local TCP port over HyperDHT (P2P)")]
struct Args {
    /// Local TCP port to expose
    #[arg(long, value_name = "PORT")]
    live: u16,

    /// Local bind address
    #[arg(long, default_value = "127.0.0.1")]
    host: String,

    /// 64-char hex seed for deterministic identity
    #[arg(long, value_name = "HEX")]
    seed: Option<String>,

    /// Reject peers that don't share the seed-derived identity.
    ///
    /// In this mode the connection key advertised to peers is the
    /// seed (not the pubkey), so only seed-holders can derive the
    /// right pubkey to dial.
    #[arg(long)]
    secure: bool,
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args = Args::parse();
    let local_addr: SocketAddr = format!("{}:{}", args.host, args.live).parse()?;

    // Pre-flight: warn if nothing's listening locally yet.
    match TcpStream::connect(&local_addr).await {
        Ok(s) => drop(s),
        Err(e) => eprintln!(
            "Warning: nothing listening on {} yet ({}). Will retry per-connection.",
            local_addr, e
        ),
    }

    let (kp, connection_key, seed_for_dht) = make_keypair(args.seed.as_deref(), args.secure)?;
    let pubkey = kp.public();

    // Build the DHT
    let dht = Arc::new(
        Dht::new(DhtOptions {
            use_public_bootstrap: true,
            // Use the same seed for the DHT identity so reconnects
            // from the same seed see a stable node identity.
            seed: seed_for_dht,
            ..Default::default()
        })
        .await?,
    );
    let dht_port = dht.port();

    // Publish JSON metadata so clients can discover the listener
    // shape (host/port/protocol). Mirrors JS holesail's mutable_put.
    // Done BEFORE listen() because listen() consumes the keypair.
    let metadata = format!(
        r#"{{"host":"{}","port":{},"udp":false}}"#,
        args.host, args.live
    );
    if let Err(e) = dht.mutable_put(&kp, metadata.as_bytes(), 1).await {
        eprintln!("Warning: mutable_put metadata failed ({}). Discovery still works via the pubkey alone.", e);
    }

    // Server with reusable_socket so reconnects from the same client
    // skip a fresh holepunch.
    let server_opts = ServerOptions {
        reusable_socket: true,
        share_local_address: true,
    };
    let mut server = dht.listen(kp, server_opts).await?;

    // Optional firewall — matches the Python version's behavior.
    // Note: the python rule (`remote_pk != self_pk`) doesn't actually
    // gate on seed-knowledge; the secrecy of "secure mode" comes from
    // the seed being kept private (it's what gets advertised, so only
    // seed-holders can derive the dial-able pubkey).
    if args.secure {
        let server_pk = pubkey;
        if let Err(e) = server.set_firewall(move |remote_pk, _host, _port| {
            remote_pk.as_bytes() != server_pk.as_bytes()
        }) {
            eprintln!("Warning: set_firewall failed ({})", e);
        }
    }

    let connection_count = Arc::new(AtomicU64::new(0));
    let active_bridges = Arc::new(AtomicU64::new(0));

    // Connection string — z32-encoded with 4-char prefix.
    let prefix = if args.secure { "s000" } else { "0000" };
    let hs_link = format!("hs://{}{}", prefix, connection_key);

    println!(
        r#"
  holesail-rs -- P2P tunnel powered by hyperdht-cpp
  --------------------------------------------------

  Exposing:    {host}:{port}
  DHT port:    {dht_port}
  Mode:        {mode}
  Public key:  {pubkey}

  Connection string:
    {hs_link}

  Connect with:
    holesail --connect {hs_link}

  Ctrl+C to stop
"#,
        host = args.host,
        port = args.live,
        dht_port = dht_port,
        mode = if args.secure { "secure" } else { "open" },
        pubkey = pubkey,
        hs_link = hs_link,
    );

    // Spawn the heartbeat task — prints state every 30s for diagnostics.
    let dht_for_heartbeat = dht.clone();
    let conns_for_heartbeat = connection_count.clone();
    let bridges_for_heartbeat = active_bridges.clone();
    let heartbeat = tokio::spawn(async move {
        let mut interval = tokio::time::interval(Duration::from_secs(30));
        // Skip the immediate first tick.
        interval.tick().await;
        loop {
            interval.tick().await;
            let addr = match dht_for_heartbeat.remote_address().await {
                Ok(Some((h, p))) => format!("{}:{}", h, p),
                _ => "unknown".to_string(),
            };
            println!(
                "  [heartbeat] online={} persistent={} addr={} bridges={} conns={}",
                dht_for_heartbeat.is_online(),
                dht_for_heartbeat.is_persistent(),
                addr,
                bridges_for_heartbeat.load(Ordering::Relaxed),
                conns_for_heartbeat.load(Ordering::Relaxed),
            );
        }
    });

    // Accept loop: each accepted Stream becomes a bidirectional bridge
    // to a freshly-opened TCP connection on `local_addr`.
    let local_addr_arc = Arc::new(local_addr);
    let accept_task = {
        let local_addr = local_addr_arc.clone();
        let connection_count = connection_count.clone();
        let active_bridges = active_bridges.clone();
        async move {
            while let Some(stream) = server.accept().await {
                let conn_id = connection_count.fetch_add(1, Ordering::Relaxed) + 1;
                let local_addr = *local_addr;
                let active_bridges = active_bridges.clone();

                tokio::spawn(async move {
                    active_bridges.fetch_add(1, Ordering::Relaxed);
                    let result = bridge_one(stream, local_addr, conn_id).await;
                    active_bridges.fetch_sub(1, Ordering::Relaxed);
                    if let Err(e) = result {
                        eprintln!("  [{}] bridge ended with error: {}", conn_id, e);
                    } else {
                        println!("  [{}] bridge closed cleanly", conn_id);
                    }
                });
            }
        }
    };

    tokio::select! {
        _ = accept_task => {}
        _ = tokio::signal::ctrl_c() => {
            println!(
                "\n  Shutting down ({} connections served)",
                connection_count.load(Ordering::Relaxed)
            );
        }
    }

    heartbeat.abort();
    // Wait for the aborted task to actually finish so it drops its
    // Arc<Dht> clone before we try_unwrap.
    let _ = heartbeat.await;

    // Drop the DHT — destroy() awaits libuv teardown gracefully.
    let dht = Arc::try_unwrap(dht)
        .map_err(|_| "dht still has outstanding references at shutdown")?;
    dht.destroy().await?;

    Ok(())
}

/// Bridge one P2P stream ↔ one local TCP connection.
async fn bridge_one(
    mut stream: hyperdht::Stream,
    local_addr: SocketAddr,
    conn_id: u64,
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    println!("  [{}] peer connected — opening local TCP", conn_id);

    let mut tcp = TcpStream::connect(local_addr).await.map_err(|e| {
        format!("can't reach {}: {}", local_addr, e)
    })?;

    // tokio handles half-close automatically: when one side EOFs,
    // copy_bidirectional shuts down the other side's writer and
    // returns once both halves are done.
    let (from_tcp, from_stream) =
        tokio::io::copy_bidirectional(&mut tcp, &mut stream).await?;
    println!(
        "  [{}] tcp→stream {} bytes, stream→tcp {} bytes",
        conn_id, from_tcp, from_stream
    );

    Ok(())
}

/// Build the keypair and the public connection key per the python
/// rules:
///
/// |  flags          | keypair source            | advertised key   |
/// | --------------- | ------------------------- | ---------------- |
/// | (none)          | random                    | pubkey           |
/// | --seed          | from seed                 | pubkey           |
/// | --secure (only) | random seed → from seed   | seed (random)    |
/// | --seed --secure | from seed                 | seed             |
///
/// Returns `(keypair, z32-encoded-connection-key, dht-seed)`. The DHT
/// seed mirrors the keypair seed (when one was provided) so the DHT
/// node identity is also deterministic across restarts.
fn make_keypair(
    seed_hex: Option<&str>,
    secure: bool,
) -> Result<(Keypair, String, Option<[u8; 32]>), Box<dyn std::error::Error>> {
    if let Some(hex_str) = seed_hex {
        if hex_str.len() != 64 {
            return Err("seed must be 64 hex characters (32 bytes)".into());
        }
        let mut seed = [0u8; 32];
        hex::decode_to_slice(hex_str, &mut seed)?;
        let kp = Keypair::from_seed(&seed);
        let pubkey_bytes = *kp.public().as_bytes();
        let advertised = if secure { seed } else { pubkey_bytes };
        Ok((kp, to_z32(&advertised), Some(seed)))
    } else if secure {
        // Generate a fresh seed; advertise the seed (it's the secret
        // that gates connectivity).
        let mut raw_seed = [0u8; 32];
        getrandom_seed(&mut raw_seed);
        let kp = Keypair::from_seed(&raw_seed);
        let advertised = raw_seed;
        Ok((kp, to_z32(&advertised), Some(raw_seed)))
    } else {
        // Open + random — advertise just the pubkey.
        let kp = Keypair::generate();
        let pubkey_bytes = *kp.public().as_bytes();
        Ok((kp, to_z32(&pubkey_bytes), None))
    }
}

/// Fill `out` with cryptographically-secure random bytes from the OS.
/// We avoid pulling in the `rand` crate for one call — this matches
/// the python version's use of `os.urandom`.
fn getrandom_seed(out: &mut [u8; 32]) {
    use std::fs::File;
    use std::io::Read;
    let mut f = File::open("/dev/urandom").expect("open /dev/urandom");
    f.read_exact(out).expect("read /dev/urandom");
}

/// Encode bytes with the z-base-32 alphabet used by holesail's
/// connection strings (`ybndrfg8ejkmcpqxot1uwisza345h769`, all
/// lowercase). 32 bytes → 52 chars.
fn to_z32(data: &[u8]) -> String {
    const Z32: &[u8] = b"ybndrfg8ejkmcpqxot1uwisza345h769";
    let mut out = String::with_capacity((data.len() * 8 + 4) / 5);
    let mut buf: u32 = 0;
    let mut bits: u32 = 0;
    for &byte in data {
        buf = (buf << 8) | byte as u32;
        bits += 8;
        while bits >= 5 {
            bits -= 5;
            let idx = ((buf >> bits) & 0x1F) as usize;
            out.push(Z32[idx] as char);
        }
    }
    if bits > 0 {
        let idx = ((buf << (5 - bits)) & 0x1F) as usize;
        out.push(Z32[idx] as char);
    }
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn z32_round_trip_known_vector() {
        // The all-zero 32-byte input under the z32 alphabet starts
        // with the alphabet's first char ('y') repeated.
        let zeros = [0u8; 32];
        let encoded = to_z32(&zeros);
        assert_eq!(encoded.len(), 52);
        assert!(encoded.chars().all(|c| c == 'y'));
    }

    #[test]
    fn z32_known_alphabet() {
        // Encoding 0x00, 0x01, 0x02 should produce known prefix
        // chars from the z32 alphabet (positions 0, 0, 1 across
        // the bit boundaries).
        let encoded = to_z32(&[0x00, 0x44, 0x32]);
        // 3 bytes = 24 bits = 5 base32 chars (last with padding bits).
        assert_eq!(encoded.len(), 5);
    }
}
